server/src/main/java/io/druid/server/security/PreResponseAuthorizationCheckFilter.java

```diff
@@ -0,0 +1,154 @@
+import io.druid.java.util.common.logger.Logger;
+import io.druid.query.QueryInterruptedException;
+import io.druid.server.DruidNode;
+import org.eclipse.jetty.server.Response;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Filter that verifies that authorization checks were applied to an HTTP request, before sending a response.
+ * <p>
+ * This filter is intended to help catch missing authorization checks arising from bugs/design omissions.
+ */
+public class PreResponseAuthorizationCheckFilter implements Filter
+{
+  private static final Logger log = new Logger(PreResponseAuthorizationCheckFilter.class);
+
+  private final AuthConfig authConfig;
+  private final List<Authenticator> authenticators;
+  private final ObjectMapper jsonMapper;
+
+  public PreResponseAuthorizationCheckFilter(
+      AuthConfig authConfig,
+      List<Authenticator> authenticators,
+      ObjectMapper jsonMapper
+  )
+  {
+    this.authConfig = authConfig;
+    this.authenticators = authenticators;
+    this.jsonMapper = jsonMapper;
+  }
+
+  @Override
+  public void init(FilterConfig filterConfig) throws ServletException
+  {
+
+  }
+
+  @Override
+  public void doFilter(
+      ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain
+  ) throws IOException, ServletException
+  {
+    if (authConfig.isEnabled()) {
+      QueryInterruptedException unauthorizedError = new QueryInterruptedException(
+          QueryInterruptedException.UNAUTHORIZED,
+          null,
+          null,
+          DruidNode.getDefaultHost()
+      );
+      unauthorizedError.setStackTrace(new StackTraceElement[0]);
+      OutputStream out = servletResponse.getOutputStream();
+
+      Boolean authInfoChecked = null;
+      final HttpServletResponse response = (HttpServletResponse) servletResponse;
+
+      // Since this is the last filter in the chain, some previous authentication filter
+      // should have placed an auth token in the request.
+      // If not, send an auth challenge.
+      if (servletRequest.getAttribute(AuthConfig.DRUID_AUTH_TOKEN) == null) {
+        Set<String> supportedAuthSchemes = Sets.newHashSet();
+        for (Authenticator authenticator : authenticators) {
+          String challengeHeader = authenticator.getAuthChallengeHeader();
+          if (challengeHeader != null) {
+            supportedAuthSchemes.add(challengeHeader);
+          }
+        }
+        for (String authScheme : supportedAuthSchemes) {
+          response.addHeader("WWW-Authenticate", authScheme);
```

Comment:
It's needed for Kerberos. the existing Kerberos auth extension does it, and it's a pretty simple method to implement (either the implementation has a sensible WWW-Authenticate scheme or it can just return null)